---
title: "Mitochondrial library report"
author: "Sergi Beneyto-Calabuig"
date: "`r format(Sys.time(), '%B %d, %Y')`"
params:
  rmd: "summary_report.Rmd"
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_float:
      collapsed: false
    highlight: pygments 
editor_options: 
  chunk_output_type: console
---

<br>


```{r, include=FALSE}

message("Checking if required packages are installed")
for (i in c("optparse","BiocManager", "GenomicRanges", "tidyverse", "knitr", "patchwork", "devtools",
            "vcfR", "RColorBrewer", "viridis", "DT", "SummarizedExperiment")){
  
  if(!requireNamespace(i, quietly = TRUE)){
  
    
    if(i %in% c("optparse", "BiocManager", "tidyverse", "RColorBrewer", "viridis",
                "DT", "vcfR", "knitr", "patchwork", "devtools")){
      
      install.packages(i, repos="https://ftp.fau.de/cran/")
      
    }else{BiocManager::install(i)}
    
  }
  
}

# set global chunk settings
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)

#load the required packages
library(tidyverse)
library(patchwork)
library(GenomicRanges)
library(vcfR)
library(RColorBrewer)
library(viridis)
library(devtools)
library(DT)
library(SummarizedExperiment)
snakemake@source("themes_mito_library.R")
snakemake@source("functions_variant_calling.R")


# get the directory where Snakefile is stored
directory <- gsub("/[^/]*$", "", snakemake@scriptdir)


# directory where plots will be exported
plots_dir <- file.path(directory, "results/summary_reports", snakemake@wildcards$patient, "plots")


# create directory to store plots
dir.create(plots_dir, recursive = T)


# the name of the sample from which the report is made
sample_report <- snakemake@wildcards$patient

# get barcodes from gene expression library
barcodes <- read_tsv(list.files(paste0(directory, "/raw_data/", sample_report),
              pattern = "barcodes.tsv", full.names = T), col_names = F) %>% pull(X1)

# get summarized experiment object with coverage information
coverage_data <- list.files(paste0(directory, "/data/", sample_report, "/mito_counts/count_table"),
             pattern = paste0("sum_experiment.rds"), full.names = T)


# get reads_cell file
reads_cell <- list.files(paste0(directory, "/data/", sample_report, "/mito_counts"),
                    pattern = "*reads_cells*", full.names = T)


# get alignment stats
align_stats <- list.files(paste0(directory, "/data/", sample_report, "/align_reads"),
               pattern = "star.Log", full.names = T)

# load data from 10x default and reference mitochondrial library (P3)
#ref_data <- readRDS(url(""))

ref_data <- readRDS("/home/sbeneyto/cluster/project/AML/git/CloneTracer/library_processing/mitochondria/data/ref_data.rds")


```


This is the summary report for the single-cell mitochondrial library of sample **`r snakemake@wildcards$patient`**.

# Proportion of reads to chrM


Here the proportion of reads aligning to the mitochondrial genome is shown. The data shown is aligned to the mitochondrial genome only in order to increase the number of mitochondrial reads. For this reason, there is no information about where the rest of the reads align (the aligner considers them unmapped since the reference genome only contains chrM).


```{r reads_mito}

# get percentage of aligned reads from star summary report
aligned_reads <- as.double(gsub(".+\t", "",read_lines(align_stats)[[9]]))

# get total number of reads
total_reads <- as.double(gsub(".+\t", "",read_lines(align_stats)[[6]]))

table <- data.frame(type = c("mitochondria", "other"),
                    nreads = c(aligned_reads, total_reads-aligned_reads)) %>%
                    mutate(percentage = nreads/sum(nreads)*100,
                    sample = sample_report,
                    type = factor(type, levels = c("other", "mitochondria")))

# put data into a dataframe
align_stats <- bind_rows(table, ref_data$align_stats) %>% 
                mutate(sample = factor(sample, levels = c(sample_report, "example_library(P3)")))


# plot with alignment locations
align_locations_plot <- ggplot(align_stats,
                           aes(x = sample, y = nreads, fill = type)) +
                            geom_bar(stat = "identity", position = "stack") +
                            theme_plots_alignment +
                            scale_y_continuous(expand = c(0,1)) +
                            scale_fill_manual(values = c(other_colour, mitochondria_purple)) +
                            geom_text(data = align_stats %>% filter(percentage > 5),
                                      aes(label = round(percentage,1)), colour = "white",
                                      position = position_stack(vjust = 0.5))  +
                            ylab("Total number of reads")


# Save the plot
ggsave(align_locations_plot + theme_plots_alignment_pdf,
       filename = file.path(plots_dir, "alignment_positions.png"),
       height = 8, width = 7)


align_locations_plot

```

<br>

***

# Coverage mitochondria

To asses the quality of the library the coverage in the mtDNA is shown. Normalised coverage indicates the mean number of UMIs per position divided by the number of reads/cell. It is difficult to compare MutaSeq v1 and v2 since in the former all reads are considered and in the latter duplicated reads are collapsed by UMI.

## Normalised coverage

<br>

```{r normalised coverage}

# load total reads/cell for all cell barcodes (unfiltered)
reads_cell <- read_csv(reads_cell, col_names = F) %>%
                separate("X1", into = c("total_reads", "cell_barcode"), sep = " ") %>%
                filter(cell_barcode %in% barcodes) %>% 
                mutate(total_reads = as.integer(total_reads),
                       sample = sample_report,
                       cell_barcode = gsub("CB:Z:(.+)$", "\\1", cell_barcode))


# convert coverage table to data frame and tidy
mito_object <- readRDS(coverage_data)

mito_object <- assays(mito_object)$nUMIs %>% as.matrix() %>%
                as.data.frame() %>%
                rownames_to_column(var = "position") %>%
                pivot_longer(names_to = "cell_barcode", values_to = "raw_coverage",
                             cols = -position) %>%
                filter(cell_barcode %in% barcodes) %>% 
                left_join(reads_cell) %>%
                mutate(norm_coverage = raw_coverage/total_reads) %>%
                group_by(position) %>%
                dplyr::summarise(mean_coverage = mean(norm_coverage, na.rm = T),
                                 raw_coverage = mean(raw_coverage, na.rm = T),
                          .groups = "keep") %>%
                ungroup() %>%
                mutate(sample = sample_report,
                       position = as.integer(position)) %>%
                filter(is.na(position) == F) %>%
                arrange(position) %>%
                mutate(smooth_coverage = runmed(mean_coverage, 31),
                       raw_coverage = runmed(raw_coverage, 31)) %>%
                dplyr::select(position, raw_coverage, mean_coverage, smooth_coverage, sample)


# bind reference and sample coverage tables
mito_coverage_norm <- bind_rows(ref_data$cvg_10x,
                                ref_data$cvg_mito_P3,
                                mito_object) %>% dplyr::select(-raw_coverage) %>%
                                mutate(sample = factor(sample))


# get the max coverage
max_norm_cvg <- mito_coverage_norm %>% pull(smooth_coverage) %>% max() + 0.01


# plot norm coverage compared to example mitochondrial library from P3
norm_coverage_plot <- ggplot(mito_coverage_norm %>% filter(sample != "10x default"),
                            aes(x = position, y = smooth_coverage, colour = sample)) +
                            geom_line(size = 1.05)  +
                            scale_color_manual(values = setNames(c(google_blue, google_red),
                                                                 c(sample_report, "example_library(P3)")))+
                            geom_segment(data = ref_data$mt_genes,
                                         aes(x = start, xend = end,
                                             y = max_norm_cvg, yend = max_norm_cvg),
                                           size = 2, colour = viridis(15, begin = 0.2)) +
                            geom_segment(data = ref_data$mito_primers,
                                        aes(x = start, xend = end,
                                           y = max_norm_cvg, yend = max_norm_cvg),
                                           size = 2, colour = "black") +
                            coord_polar() +
                            ylab("Normalised coverage") +
                            xlab("Position in the genome") +
                            theme_norm_coverage


# save the plot
ggsave(norm_coverage_plot + theme_norm_coverage_pdf,
       filename = file.path(plots_dir, "norm_coverage_mutaseqv1.png"),
       height = 7, width = 7)


norm_coverage_plot

```


***

## Raw coverage

<br>

```{r raw coverage}

# make table with raw coverage
mito_coverage_raw <- bind_rows(ref_data$cvg_mito_P3,
                               mito_object %>% dplyr::select(position, raw_coverage, sample))


# get the max coverage
max_raw_cvg <- mito_coverage_raw %>% pull(raw_coverage) %>% max() + 1500


# plot raw coverage
raw_coverage_plot <- ggplot(mito_coverage_raw,
                            aes(x = position, y = raw_coverage, colour = sample)) +
                            geom_line(size = 1.05) +
                            scale_color_manual(values = setNames(c(google_red, google_blue), 
                                                                 c("example_library(P3)",sample_report))) +
                            geom_segment(data = ref_data$mt_genes,
                                         aes(x = start, xend = end,
                                             y = max_raw_cvg, yend = max_raw_cvg),
                                           size = 2, colour = viridis(15, begin = 0.2)) +
                            geom_segment(data = ref_data$mito_primers,
                                        aes(x = start, xend = end,
                                           y = max_raw_cvg, yend = max_raw_cvg),
                                           size = 2, colour = "black") +
                            coord_polar() +
                            ylab("Raw coverage (Number of UMIs)") +
                            xlab("Position in the genome") +
                            theme_raw_coverage


# save the plot
ggsave(raw_coverage_plot + theme_raw_coverage_pdf,
       filename = file.path(plots_dir, "raw_coverage_mutaseqv1.png"),
       height = 7, width = 7)

raw_coverage_plot

```

***

## Reads/UMI

In the analysis duplicated reads with the same UMI and end position were collapsed into a consensus sequence using [CallMolecularConsensusReads](http://fulcrumgenomics.github.io/fgbio/tools/latest/CallMolecularConsensusReads.html) from fgbio. Here I plot the average reads/UMI for the mtDNA. 

```{r reads/UMI}

mito_object <- readRDS(coverage_data)

avg_reads <- mito_object@assays@data$reads_umi %>% as.matrix() %>% 
                  as.data.frame() %>% 
                  rownames_to_column(var = "position") %>%
                  pivot_longer(names_to = "cell_barcode", values_to = "nreads",
                               cols = -position) %>%
                  filter(cell_barcode %in% barcodes) %>% 
                  left_join(reads_cell) %>%
                  mutate(norm_avg_reads = nreads/total_reads) %>%
                  group_by(position) %>%
                  dplyr::summarise(norm_coverage = mean(norm_avg_reads, na.rm = T),
                                   raw_coverage = mean(nreads, na.rm = T),
                            .groups = "keep") %>%
                  ungroup() %>%
                  mutate(sample = sample_report,
                         position = as.integer(position)) %>%
                  filter(is.na(position) == F) %>%
                  arrange(position) %>%
                  mutate(smooth_coverage = runmed(norm_coverage, 31),
                         raw_coverage = runmed(raw_coverage, 31)) %>%
                  dplyr::select(position, raw_coverage, norm_coverage, smooth_coverage, sample)   

reads_umi <- bind_rows(avg_reads, ref_data$reads_umi)
  
# get the max coverage
max_umi_cvg <- reads_umi %>% pull(raw_coverage) %>% max() + 30

# plot raw coverage
raw_umi_cvg_plot <- ggplot(reads_umi,
                            aes(x = position, y = raw_coverage, colour = sample)) +
                            geom_line(size = 1.05) +
                            scale_color_manual(values = setNames(c(google_red, google_blue), 
                                                                 c("example_library(P3)",sample_report))) +
                            geom_segment(data = ref_data$mt_genes,
                                         aes(x = start, xend = end,
                                             y = max_umi_cvg, yend = max_umi_cvg),
                                             size = 2, colour = viridis(15, begin = 0.2)) +
                            geom_segment(data = ref_data$mito_primers,
                                        aes(x = start, xend = end,
                                           y = max_umi_cvg, yend = max_umi_cvg),
                                           size = 2, colour = "black") +
                            coord_polar() +
                            ylab("Average reads/UMI") +
                            xlab("Position in the genome") +
                            theme_umi_coverage


# save the plot
ggsave(raw_umi_cvg_plot + theme_umi_coverage_pdf,
       filename = file.path(plots_dir, "reads_umi_sample.png"),
       height = 7, width = 7)


raw_umi_cvg_plot



```

***

## Reads/amplicon

The snakemake pipeline also extracts the number of reads supporting each amplicon. Here I make a histogram of the values. Amplicons with > 10 reads were set to 10 reads. This plot is useful to determine the sequencing saturation levels.

<br>

```{r reads amplicon}

# read file with reads/amplicon
reads_amplicon <- read_delim(paste0(directory,
                                    "/data/", sample_report, "/mito_counts/count_table/nreads_umi.txt"),
                             delim = "\t", col_names = F) %>% 
                              mutate(nreads = ifelse(X1 > 10, 10, X1))

# make histogram
hist_reads_amplicon <- ggplot(reads_amplicon, 
                              aes(x = nreads))+
                              geom_histogram(bins = 9) +
                              ylab("Number of amplicons") +
                              theme_classic() +
                              scale_x_continuous(breaks = 1:10) +
                              scale_y_continuous(expand = c(0,0)) +
                              theme(axis.title = element_text(size = 14),
                                    axis.text = element_text(size = 13))

ggsave(hist_reads_amplicon,
       filename = file.path(plots_dir, "hist_reads_umi.png"),
       height = 7, width = 7)


hist_reads_amplicon

```

***

## Base quality

Here I show the average base quality/cell over the mitochondrial genome.

```{r base quality}

qual_list <- lapply(1:length(c("A_qual", "T_qual", "C_qual", "G_qual")), function(j){
  
  nt <- c("A_qual", "T_qual", "C_qual", "G_qual")[j]
  
  mean_qual <- mito_object@assays@data[[nt]] %>% as.matrix() %>% 
                  as.data.frame() %>% 
                  rownames_to_column(var = "position") %>%
                  pivot_longer(names_to = "cell_barcode", values_to = "qual",
                               cols = -position) %>%
                  filter(cell_barcode %in% barcodes) %>% 
                  filter(qual > 0) %>% 
                  group_by(position) %>%
                  dplyr::summarise(avg_qual = mean(qual, na.rm = T),
                            .groups = "keep") %>%
                  ungroup() %>%
                  mutate(position = as.integer(position)) %>%
                  filter(is.na(position) == F) %>%
                  arrange(position) 
  
  colnames(mean_qual) <- c("position", nt)
  
  mean_qual
})

qual_table <- tibble(position = 1:16569) %>% 
                    left_join(qual_list[[1]]) %>% 
                    left_join(qual_list[[2]]) %>% 
                    left_join(qual_list[[3]]) %>% 
                    left_join(qual_list[[4]]) %>% 
                    pivot_longer(cols = -position, 
                                 names_to = "nucleotide", values_to = "qual") %>% 
                    group_by(position) %>% 
                    dplyr::summarise(avg_qual = max(qual, na.rm = T),
                                     .groups = "keep") %>% 
                    ungroup() %>% 
                    mutate(avg_qual = ifelse(is.nan(avg_qual) | is.na(avg_qual), 0, avg_qual),
                           sample = sample_report,
                           avg_qual = runmed(avg_qual, 31)) 

base_qual_table <- bind_rows(qual_table, ref_data$base_qual)

# get the max quality
max_qual <- base_qual_table %>% pull(avg_qual) %>% max() + 5

# plot raw coverage
qual_plot <- ggplot(base_qual_table,
                      aes(x = position, y = avg_qual, colour = sample)) +
                      geom_line(size = 1.05) +
                      scale_color_manual(values = setNames(c(google_blue, google_red), 
                                                           c(sample_report, "example_library(P3)"))) +
                      geom_segment(data = ref_data$mt_genes,
                                   aes(x = start, xend = end,
                                       y = max_qual, yend = max_qual),
                                       size = 2, colour = viridis(15, begin = 0.2)) +
                      geom_segment(data = ref_data$mito_primers,
                                  aes(x = start, xend = end,
                                     y = max_qual, yend = max_qual),
                                     size = 2, colour = "black") +
                      coord_polar() +
                      ylab("Average base quality") +
                      xlab("Position in the genome") + 
                      theme_base_qual


# save the plot
ggsave(qual_plot + theme_base_qual_pdf,
       filename = file.path(plots_dir, "base_quality_sample.png"),
       height = 7, width = 7)


qual_plot

```


# Source
* <a download="summary_report.Rmd" href="`r base64enc::dataURI(file = params$rmd, mime = 'text/rmd',
    encoding = 'base64')`">R Markdown source file (to produce this document)</a>

