# Description

The R script [design_primers.R](design_primers) is used to design primers for the amplification of specific nuclear SNVs from 3' 10x scRNAseq cDNA libraries. For eac mutation it generates outer, middle and staggered inner PCRs which can be used following the protocol described in out MutaSeq v2 manuscrip.

## Dependencies 

In order to run the script, samtools is required in the environment as well as the following R packages: tidyverse, GenomicRanges, GenomicFeatures, Seurat, BiocParallel, rtracklayer, BSgenome, TAPseq, ballgown, purrr, TxDb.Hsapiens.UCSC.hg38.knownGene, mygene, BSgenome.Hsapiens.UCSC.hg38.

We have provide a conda environment with all required dependencies. It can be installed with the following code:

```
git clone https://github.com/veltenlab/MutaSeq-v2
cd primer_design
conda env create -f envs/primer_design.yml
conda activate primer_design
```

## Design primers

### Considerations

There are 2 main limitations when it comes to genotyping nuclear SNVs: the expression of the gene and the distance of the mutation to the polyA tail (end of mRNA). By default we select mutations that are < 1.5kb away from the estimated polyA and genes which are expressed in the sample of interest (mean counts/cell > 0.2).

From our experience mutations in genes which have lower expression is highly inefficient (only a low proportion of cells is covered). When it comes to mutations further away from the polyA tail, the main issue is the length of the resulting fragments. We have noticed that longer fragments (>1kb) are sequenced less efficiently than shorter ones. Furthermore, amplifying very long fragments can be problematic when carrying out the nested PCRs.

### Input files

There are 3 main input files required to run the script:

* csv file with the following format:

| Symbol      | CHROM  | POS
| ----------- | ------ |-----------
| KRAS        | chr12  | 25245350
| NRAS        | chr1   | 114713909
| IDH2        | chr15  | 90088702

* BAM file from 3' 10x scRNAseq of the same sample
* [gtf file from ensembl](http://ftp.ensembl.org/pub/release-100/gtf/homo_sapiens/Homo_sapiens.GRCh38.100.chr.gtf.gz).

### Command-line execution

The script has the following command-line arguments:

#### Required

* `-i, --input <filepath>`: input csv file with the format described above
* `-b --bam <filepath>`: bam file from 3' 10x genomics
* `-u --outdir_gene_exp <directory>`: outs directory generated by cellranger pipeline. It should include directory filtered_feature_bc_matrix used to create a Seurat object
* `-g --gtf_file <filepath>`: gtf file from ensembl
* `-n --name <text>`: sample name
* `-r --read_length <integer>`: read length of read2 
* `-d --out_directory <directory>`: directory where output files will be written
* `-t --multimodal <logical>`: TRUE when additional modalities where measured besided RNA (e.g. surface antigens, ATAC...)

#### Optional

* `-c --cores <integer>`: number of cores to use. 
* `-m --forced_mutations <filepath>`: text file with tab-separated gene names for which primers will be designed regardless of the expression of the gene or the distance of the mutation to the polyA.

## Output

The script produces 2 main output files

```
*/primers/primer_sequence.csv
*/primers/primer_details.csv
```

The former contains the name of the primers as well as the sequence. The latter contains more detailed information about each primer (e.g. temperature of melting, distance from the mutation of interest...)

The script generates several bed files stored in the `genomic_files` and in `primers` directories which can be loaded into a genome browser to confirm the correct orientation and position of the primers. 

In our experience visualising these files has been helpful particularly in instances where the initial PCR reactions did not seem to work.
